lightning-invoice/src/payment.rs

```diff
@@ -244,140 +253,166 @@ where
-					hash_map::Entry::Vacant(entry) => entry.insert(0),
-				};
-			};
-			if let hash_map::Entry::Occupied(mut entry) = entry {
-				let max_payment_attempts = self.retry_attempts.0 + 1;
-				let attempts = entry.get_mut();
-				*attempts += 1;
-
-				if *attempts >= max_payment_attempts {
-					log_trace!(self.logger, "Payment {} exceeded maximum attempts; not retrying (attempts: {})", log_bytes!(payment_hash.0), attempts);
-					return Err(());
-				} else if has_expired(params) {
-					log_trace!(self.logger, "Invoice expired for payment {}; not retrying (attempts: {})", log_bytes!(payment_hash.0), attempts);
-					return Err(());
-				}
+	fn retry_payment(
+		&self, payment_id: PaymentId, payment_hash: PaymentHash, params: &RouteParameters
+	) -> Result<(), ()> {
+		let max_payment_attempts = self.retry_attempts.0 + 1;
+		let attempts = *self.payment_cache.lock().unwrap()
+			.entry(payment_hash)
+			.and_modify(|attempts| *attempts += 1)
+			.or_insert(1);
+
+		if attempts >= max_payment_attempts {
+			log_trace!(self.logger, "Payment {} exceeded maximum attempts; not retrying (attempts: {})", log_bytes!(payment_hash.0), attempts);
+			return Err(());
+		}
 
-				let payer = self.payer.node_id();
-				let first_hops = self.payer.first_hops();
-				route = self.router.find_route(&payer, &params, Some(&first_hops.iter().collect::<Vec<_>>()), &self.scorer.lock());
-				if route.is_err() {
-					log_trace!(self.logger, "Failed to find a route for payment {}; not retrying (attempts: {})", log_bytes!(payment_hash.0), attempts);
-					return Err(());
-				}
-			} else {
-				unreachable!();
-			}
+		if has_expired(params) {
+			log_trace!(self.logger, "Invoice expired for payment {}; not retrying (attempts: {})", log_bytes!(payment_hash.0), attempts);
+			return Err(());
 		}
 
-		let retry_res = self.payer.retry_payment(&route.unwrap(), payment_id);
-		match retry_res {
+		let payer = self.payer.node_id();
+		let first_hops = self.payer.first_hops();
+		let route = self.router.find_route(
+			&payer, &params, &payment_hash, Some(&first_hops.iter().collect::<Vec<_>>()),
+			&self.scorer.lock()
+		);
+		if route.is_err() {
+			log_trace!(self.logger, "Failed to find a route for payment {}; not retrying (attempts: {})", log_bytes!(payment_hash.0), attempts);
+			return Err(());
+		}
+
+		match self.payer.retry_payment(&route.unwrap(), payment_id) {
 			Ok(()) => Ok(()),
 			Err(PaymentSendFailure::ParameterError(_)) |
 			Err(PaymentSendFailure::PathParameterError(_)) => {
 				log_trace!(self.logger, "Failed to retry for payment {} due to bogus route/payment data, not retrying.", log_bytes!(payment_hash.0));
-				return Err(());
+				Err(())
 			},
 			Err(PaymentSendFailure::AllFailedRetrySafe(_)) => {
 				self.retry_payment(payment_id, payment_hash, params)
 			},
-			Err(PaymentSendFailure::PartialFailure { results: _, failed_paths_retry, .. }) => {
+			Err(PaymentSendFailure::PartialFailure { failed_paths_retry, .. }) => {
```

Comment:
Do we test the `PartialFailure` send/retry_payment case?

Reply:
Looks like we don't! Added another test.